D3D11 Backend

hrydgard opened the pull request:
This will be needed for a good performance UWP port later.

Works pretty well, a few bugs left to deal with but a lot of games work so I figured it's time to let everyone play with this.

hrydgard committed:
Start stubbing out a new D3D11 backend

hrydgard committed:
Enable selecting D3D11 on Windows

hrydgard committed:
Some mostly empty files for D3D11

hrydgard committed:
Initial stab at porting the DX9 backend to D3D11. Not working yet.

hrydgard committed:
Further stuff.

hrydgard committed:
Make the first generated D3D11 shaders compile, fix various warnings

hrydgard committed:
Break on D3D11 validation errors in _DEBUG mode

hrydgard committed:
More D3D11 progress

hrydgard committed:
cube.elf now runs in D3D11 mode, still blackscreen

hrydgard committed:
D3D11: Cube now visible

hrydgard committed:
D3D11: Fix depth state

hrydgard committed:
More D3D11 improvements

hrydgard committed:
D3D11: Fix clears

hrydgard committed:
D3D11: Lumines is now playable in non-buffered, with reverse colors.

hrydgard committed:
D3D11: Fix reversed colors, stencil improvements

zminhquanz commented:
it's good for Windows 10 Mobile , hope to play PPSSPP to day by day

hrydgard commented:
Yup, that's the idea (and also Xbox One in developer mode...).  Still lots of work left to do on this of course, it's early.

hrydgard committed:
Assorted fixes for memory leaks etc

hrydgard committed:
D3D11: Change init method for textures, so we can support mipmapping

LunaMoo commented on Windows/GPU/D3D11Context.cpp:
```diff
@@ -98,6 +111,19 @@ void D3D11Context::Resize() {
+	d3dDebug_->ReportLiveDeviceObjects(D3D11_RLDO_SUMMARY | D3D11_RLDO_DETAIL);
+#endif
+	d3dDebug_->Release();
+	d3dInfoQueue_->Release();
```
Guessing those two should be included under _DEBUG above. It also seems to be repeated without reason, I mean the ifdef.

LunaMoo commented:
New backends always nice:3.

 I'm a bit confused since it crashes instantly without any error if I try to run it normally, but runs fine when I start it under VS ~ not sure how to find the reason of crash in this situation @_@

DonelBueno commented:
How will this compare to DX9 and OpenGL backends in terms of speed and accuracy?

moura464 commented:
xbox one xdk mode

hrydgard commented:
@LunaMoo  worry about that when I remove the WIP tag :) Still lots of work to do.

@DonelBueno It's an attempt at a straight port of the DX9 backend, so the initial goal is to match the DX9 backend in all respects, including performance hopefully. After that all works, it might make sense to try to make use of DX11-only functionality to gain a little more speed or features.

hrydgard committed:
Android buildfix

xebra commented:
Welcome new backend 👍 
Just build and tested on Win7, some games working fine but some games crash ;o
I know it has WIP tag. Please to keep great work!

hrydgard committed:
Fix DrawActiveTexture, now Buffered Rendering is starting to work

hrydgard committed:
Assorted D3D11 fixes

hrydgard committed:
D3D11: Fix reverse colors in drawpixels, fix a potential crash

hrydgard committed:
D3D11: Fix build issue in Release mode. Triangle fan workaround.

hrydgard commented:
With that, a lot more games should start showing stuff at least, and buffered rendering now appears to mostly work.

hrydgard committed:
Fix the stencil buffer initialization stuff (not sure it's 100% working et)

hrydgard committed:
Windows GEDebugger: Make display list view DPI-aware for easier debugging on high-dpi screens

hrydgard committed:
Further D3D11 fixes

hrydgard committed:
Minor tweaks

hrydgard committed:
D3D11: Stencil and scissor work

hrydgard committed:
D3D11: Prepare for dual src blending support

hrydgard committed:
DX9 typo fix in PackDepthBuffer

hrydgard committed:
D3D11: Some cleanup. Fix backface culling. Wipeout still looks weird...

hrydgard committed:
Additional D3D11 fixes

hrydgard committed:
More D3D11 fixes, hopefully without breaking DX9

hrydgard committed:
Last one for today

hrydgard committed:
D3D11 Attempt at fixing Star Ocean stencil .. doesn't quite work.

moura464 commented:
https://technet.microsoft.com/pt-pt/library/mt427455

moura464 commented:
https://technet.microsoft.com/en-au/library/mt427455

hrydgard committed:
D3D11: Fix some color issues by doing thing similarly to DX9. Can be optimized later.

hrydgard committed:
D3D11 Stencil ref fix, fixes Star Ocean stencil-depth

zminhquanz commented:
If you can optimize and more compabillity , i think it's better than OpenGL in Windows environment

hrydgard commented:
@zminhquanz It already works pretty well in most games and is faster than GL, and once I've fixed all the issues, it'll be the new default backend on Win8+ (and later Win7 when I've removed the reliance on a texture format that only works on Win8+). All my refactoring work before was worth it, it made it quite easy to implement.

hrydgard committed:
Hide the D3D11 backend if Windows version is older than Win8

hrydgard committed:
D3D11: Improved alpha/color test, like in GL/Vulkan

hrydgard committed:
Enable dualsource blend effects (glows on transparent things in Wipeout, for example)

hrydgard commented:
Will merge soon.

Currently missing features are:

* Framebuffer readback effects
* A few strangely missing draw calls in a few games
* Framebuffer "Depalettization" (Sonic Riders, shadows in FF:Type-1, etc).

hrydgard committed:
D3D11: Support window resize

Anuskuss commented:
@hrydgard You're a god. Hope to test it soon on Win7 :+1:

unknownbrackets commented on GPU/Common/SoftwareTransformCommon.cpp:
```diff
@@ -502,7 +502,7 @@ void SoftwareTransform(
 		u16 *newInds = inds + vertexCount;
 		u16 *indsOut = newInds;
-		maxIndex = 4 * vertexCount;
+		maxIndex = 4 * (vertexCount / 2);
```
Not 6 *?  Isn't this a count of index values?

-[Unknown]

hrydgard replied:
Nope, it's confusingly the highest index value + 1,  IIRC.

unknownbrackets replied:
Oh, duh.

-[Unknown]

unknownbrackets commented on GPU/D3D11/StencilBufferD3D11.cpp:
```diff
@@ -31,42 +31,46 @@
+};
+float4 main(PS_IN In) : SV_Target {
+  float4 index = tex.Sample(samp, In.v_texcoord0);
+	uint indexBits = uint(index.a * 256.0);
```
Won't an alpha of `1.0` become `256` and then `indexBits & 0x100` won't work?

-[Unknown]

hrydgard replied:
Oh right, that was the motivation for 255.99..

unknownbrackets commented on GPU/D3D11/FramebufferManagerD3D11.cpp:
```diff
@@ -0,0 +1,1012 @@
+	if (matchingDepthBuffer && matchingSize) {
+		// Doesn't work.  Use a shader maybe?
+		draw_->BindBackbufferAsRenderTarget();
+		draw_->CopyFramebufferImage(src->fbo, 0, 0, 0, 0, dst->fbo, 0, 0, 0, 0, src->renderWidth, src->renderHeight, 1, Draw::FB_DEPTH_BIT);
```
Doesn't this also copy stencil, which isn't necessarily desirable (since the PSP keeps stencil with the color buffer, which hasn't been switched)?  This would also apply to using StretchRect on D3D9, right?

-[Unknown]

hrydgard replied:
Yes, you're right :/ This needs further work.
